src/rust/engine/nailgun_process_map/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+//#![deny(warnings)]
+// Enable all clippy lints except for many of the pedantic ones. It's a shame this needs to be copied and pasted across crates, but there doesn't appear to be a way to include inner attributes from a common source.
+#![deny(
+clippy::all,
+clippy::default_trait_access,
+clippy::expl_impl_clone_on_copy,
+clippy::if_not_else,
+clippy::needless_continue,
+clippy::single_match_else,
+clippy::unseparated_literal_suffix,
+clippy::used_underscore_binding
+)]
+// It is often more clear to show that nothing is being moved.
+#![allow(clippy::match_ref_pats)]
+// Subjective style.
+#![allow(
+clippy::len_without_is_empty,
+clippy::redundant_field_names,
+clippy::too_many_arguments
+)]
+// Default isn't as big a deal as people seem to think it is.
+#![allow(clippy::new_without_default, clippy::new_ret_no_self)]
+// Arc<Mutex> can be more clear than needing to grok Orderings:
+#![allow(clippy::mutex_atomic)]
+
+// TODO: Maybe rename to JVMProcessMap, to decouple from the Nailgun solution?
+// Maybe make more general, to manage any long-running process.
+use bytes::Bytes;
+use std::collections::HashMap;
+use sysinfo::{Pid, SystemExt, ProcessExt};
+
+use hashing::Fingerprint;
+use process_execution::ExecuteProcessRequest;
+use std::path::{PathBuf, Path, Component};
+use std::hash::{Hash, Hasher};
+use std::fs::{metadata, File};
+use std::{fs, io};
+use std::io::{BufRead, BufReader};
+use std::io::Write;
+use std::collections::hash_map::DefaultHasher;
+use std::process::Stdio;
+use regex::Regex;
+use lazy_static::lazy_static;
+
+lazy_static! {
+    static ref NAILGUN_PORT_REGEX: Regex = Regex::new(r".*\s+port\s+(\d+)\.$").unwrap();
+}
+
+// TODO: This can be just an enum, but using an enum while developing.
+type NailgunProcessName = String;
+//type NailgunProcessFingerprint = Fingerprint;
+type NailgunProcessFingerprint = u64;
+type Port = usize;
+
+pub struct NailgunProcessMap {
+    // TODO: Possibly wrap in a Mutex
+    processes: HashMap<NailgunProcessName, NailgunProcessMetadata>,
+    system: sysinfo::System,
+}
+
+fn hacky_hash(epr: &ExecuteProcessRequest) -> NailgunProcessFingerprint {
+    // TODO Use CommandRunner.digest here!
+    let mut hasher = DefaultHasher::new();
+    epr.hash(&mut hasher);
+    hasher.finish()
+}
+
+impl NailgunProcessMap {
+    pub fn new() -> Self {
+        NailgunProcessMap {
+            processes: HashMap::new(),
+            system: sysinfo::System::new(),
+        }
+    }
+
+    pub fn connect(&mut self, name: NailgunProcessName, startup_options: ExecuteProcessRequest) -> Result<&NailgunProcessMetadata, String> {
```

Comment:
Before continuing too much in isolation like this, I'd go ahead and integrate this, for two reasons:
1. This needs to be async, using the `tokio-process` APIs we use currently
2. Your "does this need to be concurrency safe" question is easiest to answer by having the compiler tell you, rather than by trying to determine the surface area before integrating.

Reply:
Started integration in: https://github.com/pantsbuild/pants/blob/7005c70d30b1e9a046c18a0fafd05afcdd91cd5a/src/rust/engine/process_execution/src/nailgun/mod.rs
1. I have made all the signatures futuresy, but will work tomorrow on changing hte implementation. It already used Results and maps over Results, so it shouldn't be too hard.
2. Yeah, I needed the mutex for interior mutability: https://pantsbuild.slack.com/archives/C046T6T9U/p1570030622094200

Reply:
Integrated!